File: src/ainfra/inventory.py
"""Deterministic Ansible inventory generation from validated output."""

from __future__ import annotations

from pathlib import Path
from typing import Any

import yaml  # type: ignore[import-untyped]

from ainfra.contracts import validate_path
from ainfra.errors import GuardError
from ainfra.policy import validate_policy


def build_inventory(output_path: Path) -> dict[str, Any]:
    """Build a stable, secret-free inventory from an InfrastructureOutput."""

    document = validate_path(output_path)
    if document["kind"] != "InfrastructureOutput":
        raise GuardError("inventory requires an InfrastructureOutput document")
    validate_policy(document, source=output_path)

    children: dict[str, Any] = {
        "control_plane": {"hosts": {}},
        "workers": {"hosts": {}},
    }
    for node in sorted(
        document["spec"]["nodes"], key=lambda item: item["name"]
    ):
        group = (
            "control_plane"
            if node["role"] == "control-plane-capable"
            else "workers"
        )
        children[group]["hosts"][node["name"]] = {
            "ansible_host": node["privateIPv4"],
            "ainfra_image": node["image"],
            "ainfra_role": node["role"],
        }

    return {
        "all": {
            "vars": {
                "ansible_user": "ainfra",
                "ansible_python_interpreter": "/usr/bin/python3",
                "ainfra_private_cidr": document["spec"]["network"][
                    "privateCidrs"
                ][0],
            },
            "children": children,
        }
    }


def write_inventory(output_path: Path, destination: Path) -> None:
    """Write inventory atomically without emitting credentials."""

    inventory = build_inventory(output_path)
    destination.parent.mkdir(parents=True, exist_ok=True)
    temporary = destination.with_suffix(f"{destination.suffix}.tmp")
    temporary.write_text(
        yaml.safe_dump(inventory, sort_keys=False),
        encoding="utf-8",
    )
    temporary.replace(destination)

File: src/inventory.rs
//! Deterministic Ansible inventory generation.

use std::collections::BTreeMap;
use std::fs;
use std::path::Path;

use serde::Serialize;

use crate::contracts::validate_path;
use crate::error::AinfraError;
use crate::policy::validate_policy;

#[derive(Debug, Serialize)]
struct Inventory {
    all: All,
}

#[derive(Debug, Serialize)]
struct All {
    vars: Variables,
    children: Children,
}

#[derive(Debug, Serialize)]
struct Variables {
    ansible_user: &'static str,
    ansible_python_interpreter: &'static str,
    ainfra_private_cidr: String,
}

#[derive(Debug, Serialize)]
struct Children {
    control_plane: Group,
    workers: Group,
}

#[derive(Debug, Default, Serialize)]
struct Group {
    hosts: BTreeMap<String, Host>,
}

#[derive(Debug, Serialize)]
struct Host {
    #[serde(rename = "ansible_host")]
    address: String,
    #[serde(rename = "ainfra_image")]
    image: String,
    #[serde(rename = "ainfra_role")]
    role: String,
}

/// Build a validated, secret-free inventory as YAML.
///
/// # Errors
///
/// Returns a stable contract, policy, or guard error.
pub fn build_inventory(output_path: &Path, project_root: &Path) -> Result<String, AinfraError> {
    let document = validate_path(output_path)?;
    if document["kind"] != "InfrastructureOutput" {
        return Err(AinfraError::guard(
            "inventory requires an InfrastructureOutput document",
        ));
    }
    validate_policy(&document, output_path, None, project_root)?;
    let mut control_plane = Group::default();
    let mut workers = Group::default();
    let mut nodes = document["spec"]["nodes"]
        .as_array()
        .cloned()
        .unwrap_or_default();
    nodes.sort_by_key(|node| node["name"].as_str().unwrap_or_default().to_owned());
    for node in nodes {
        let name = node["name"].as_str().unwrap_or_default().to_owned();
        let host = Host {
            address: node["privateIPv4"].as_str().unwrap_or_default().to_owned(),
            image: node["image"].as_str().unwrap_or_default().to_owned(),
            role: node["role"].as_str().unwrap_or_default().to_owned(),
        };
        if node["role"] == "control-plane-capable" {
            control_plane.hosts.insert(name, host);
        } else {
            workers.hosts.insert(name, host);
        }
    }
    let private_cidr = document["spec"]["network"]["privateCidrs"][0]
        .as_str()
        .unwrap_or_default()
        .to_owned();
    serde_yaml::to_string(&Inventory {
        all: All {
            vars: Variables {
                ansible_user: "ainfra",
                ansible_python_interpreter: "/usr/bin/python3",
                ainfra_private_cidr: private_cidr,
            },
            children: Children {
                control_plane,
                workers,
            },
        },
    })
    .map_err(|error| AinfraError::dependency(error.to_string()))
}

/// Write inventory atomically through a sibling temporary file.
///
/// # Errors
///
/// Returns a stable validation or filesystem error.
pub fn write_inventory(
    output_path: &Path,
    destination: &Path,
    project_root: &Path,
) -> Result<(), AinfraError> {
    let inventory = build_inventory(output_path, project_root)?;
    if let Some(parent) = destination.parent() {
        fs::create_dir_all(parent).map_err(|error| AinfraError::dependency(error.to_string()))?;
    }
    let temporary_extension = destination.extension().map_or_else(
        || "tmp".to_owned(),
        |suffix| format!("{}.tmp", suffix.to_string_lossy()),
    );
    let temporary = destination.with_extension(temporary_extension);
    fs::write(&temporary, inventory).map_err(|error| AinfraError::dependency(error.to_string()))?;
    fs::rename(&temporary, destination).map_err(|error| AinfraError::dependency(error.to_string()))
}
